cocotb/share/lib/simulator/simulatormodule.cpp

```diff
@@ -588,6 +590,54 @@ static PyObject *get_signal_val_long(gpi_hdl_Object<gpi_sim_hdl> *self, PyObject
     return PyLong_FromLong(result);
 }
 
+static PyObject *set_signal_val_bytes(gpi_hdl_Object<gpi_sim_hdl> *self, PyObject *args)
+{
+    gpi_set_action_t action;
+    const char *bytes;
+    Py_ssize_t len;
+
+    if (!PyArg_ParseTuple(args, "iy#:set_signal_val_bytes", &action, &bytes, &len)) {
+        return NULL;
+    }
+
+    const std::array<std::string, 16> BINSTR_LUT = {
+        {"0000", "0001", "0010", "0011", "0100", "0101", "0110", "0111",
+         "1000", "1001", "1010", "1011", "1100", "1101", "1110", "1111"}
+    };
+
+    const std::array<std::string, 16> HEXSTR_LUT = {
+        {"0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "a", "b", "c", "d", "e", "f"}
+    };
+
+    int m_elems = gpi_get_num_elems(self->hdl);
+
+    std::string m_binstr{};
+    std::string m_hexstr{};
+    for (auto i = 0; i < len; ++i)
+    {
+        m_binstr += BINSTR_LUT[static_cast<int>((bytes[i] & 0xf0) >> 4)];
+        m_binstr += BINSTR_LUT[static_cast<int>( bytes[i] & 0x0f)];
+        m_hexstr += HEXSTR_LUT[static_cast<int>((bytes[i] & 0xf0) >> 4)];
+        m_hexstr += HEXSTR_LUT[static_cast<int>( bytes[i] & 0x0f)];
+    }
+
+    if (m_binstr.length() > static_cast<size_t>(m_elems)) {
+        std::string m_msb = m_binstr.substr(0, m_binstr.length() - m_elems);
+        bool m_ovfl = m_msb.find_first_not_of(m_msb[0]) != std::string::npos; // are MSBs not all the same?
+
+        if (m_ovfl) {
+            const char *m_name = gpi_get_signal_name_str(self->hdl);
+            PyErr_Format(PyExc_OverflowError, "Value (0x%s) out of range for assignment of %d-bit signal (%s)", m_hexstr.c_str(), m_elems, m_name);
+            return NULL;
+        } else {
+            m_binstr = m_binstr.substr(m_binstr.length() - m_elems, std::string::npos); // safely drop MSBs
+        }
+    }
+
+    gpi_set_signal_value_binstr(self->hdl, m_binstr.c_str(), action);
```

Comment:
We could investigate if adding gpi_set_signal_value_hexstr(), using vpiHexStrVal/vhpiHexStrVal would improve performance.